src/config.ts

```diff
@@ -20,20 +23,55 @@ const argv = yargs.options({
     }
 }).argv as CartaCommandLineOptions;
 
+const usingCustomConfig = argv.config !== defaultConfigPath;
 const testUser = argv.test;
 const configSchema = require("../config/config_schema.json");
-const ajv = new Ajv({useDefaults: true});
+const ajv = new Ajv({useDefaults: false});
+const ajvWithDefaults = new Ajv({useDefaults: true});
 const validateConfig = ajv.compile(configSchema);
+const validateAndAddDefaults = ajvWithDefaults.compile(configSchema);
 
 let serverConfig: CartaServerConfig;
 
 try {
     console.log(`Checking config file ${argv.config}`);
-    const jsonString = fs.readFileSync(argv.config).toString();
-    serverConfig = parse(jsonString);
-    const isValid = validateConfig(serverConfig);
+    if (fs.existsSync(argv.config)) {
+        const jsonString = fs.readFileSync(argv.config).toString();
+        serverConfig = JSONC.parse(jsonString);
+    } else {
+        if (!usingCustomConfig) {
+            serverConfig = {} as CartaServerConfig;
+            console.log(`Skipping missing config file ${defaultConfigPath}`);
+        } else {
+            console.log(new Error(`Unable to find config file ${argv.config}`));
+            process.exit(1);
+        }
+    }
+
+    const configDir = path.join(path.dirname(argv.config), "config.d")
+    if (fs.existsSync(configDir)) {
+        const files = fs.readdirSync(configDir);
```

Comment:
Is the order here predictable or random? The merge order may be important.

Am I correct in my understanding of the `authProviders` section? Is it a single object with one named property per possible provider type, so that you can have a `pam` and an `external` provider, but not two `external` providers, because the second will override the first? Would it not make more sense for this to be a list of provider objects which have a type property?

Reply:
the two will be deep-merged, so if you have two `pam` objects, they'll get combined. I suppose we could move to the approach you suggested, but that would be a rather large change, and we still don't know exactly how we're going to support multiple providers

Reply:
@confluence what should we do about this? Can we leave it for a later issue?

Reply:
I think the deep merge is fine for now, as long as I explain it in the docs -- although I would like to switch to something more intuitive later. I would like to confirm that the merge order is alphabetical before merging, though.